Add GRPC client tracing interceptor for grpcs when cert is given

Breina opened the pull request:
This case seems to have been mistakenly omitted? Why not trace this code path as well?

Breina committed:
Add GRPC client tracing interceptor for grpcs when cert is given

Signed-off-by: Breina <info@breina.be>

Breina commented:
Do you need me to create an Issue for this first, or how do we go about this?

bestbeforetoday reviewed (commented):
Although it's generally nice to have an associated issue, this looks like a small enough fix/change that I can live without one.

@atoulme Do you know any reason why the GrpcTelemetry client interceptor was not (or should not) be included in this code path?

Unless there is some reason not to add the interceptor for this code path, the change looks reasonable to me. Just two things that would be good to consider:

1. Is it possible to unit test the Endpoint creation to check that an interceptor is correctly added. This would avoid anyone breaking it in the future.
2. This change would make all the non-error code paths for the outer try/catch block call both `.intercept(clientInterceptor)` and `addNettyBuilderProps(channelBuilder, properties)`. Could the code be refactored so those calls happen at one common point rather than being duplicated in three code paths, which makes it easy for one to be missed.

Breina commented:
Sure, I'm down to refactor it.

Breina commented:
Alright I refactored it. I looked into unit testing the interceptor, but it's inside netty's builder and makes it complex to get it out. I've tested locally instead.

Breina committed:
Refactor Endpoint.java

- Split into digestible chunks
- Reduce cyclometer complexity
- Improve code reuse

Signed-off-by: Breina <info@breina.be>

bestbeforetoday commented on src/main/java/org/hyperledger/fabric/sdk/Endpoint.java:
```diff
@@ -61,243 +70,318 @@
+    private static Optional<byte[]> getPEMFile(String url, Properties properties) {
+        byte[] pemBytes;
+        try (ByteArrayOutputStream bis = new ByteArrayOutputStream(64000)) {
+            @SuppressWarnings("UseOfPropertiesAsHashtable") byte[] pb = (byte[]) properties.get("pemBytes");
```
It might be worth extracting `"pemBytes"` out as a constant as you have done nicely for the other property names

bestbeforetoday commented on src/main/java/org/hyperledger/fabric/sdk/Endpoint.java:
```diff
@@ -385,90 +475,122 @@ private void addNettyBuilderProps(NettyChannelBuilder channelBuilder, Properties
-            Utils.invokeMethod(channelBuilder, methodName, classParms, parmsArray);
+            try {
+                Utils.invokeMethod(channelBuilder, methodName, classParms, parmsArray);
+            } catch (IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+                logger.error(format("Endpoint %s, exception '%s'", url, e.getMessage()), e);
+                logger.error(e);
+                throw new RuntimeException(e);
+            }
```
I think this now duplicates the logging done in the Endpoint constructor. It might be simpler to declare these as thrown exceptions and have them handled by the constructor, similar to the original code

bestbeforetoday commented on src/main/java/org/hyperledger/fabric/sdk/Endpoint.java:
```diff
@@ -385,90 +475,122 @@ private void addNettyBuilderProps(NettyChannelBuilder channelBuilder, Properties
+            String propertyClientKeyFile = properties.getProperty(CLIENT_KEY_FILE);
+            String propertyClientCertFile = properties.getProperty(CLIENT_CERT_FILE);
+
+            if (propertyClientKeyFile != null && propertyClientCertFile != null) {
```
You could reverse the logic here and move the throw from the else block this if block, similar to the short-circuit return you used at the top of this method. This would avoid the need for the following code to be in a nested block

bestbeforetoday reviewed (changes requested):
Thank you for the effort you have put into this refactor. A few small suggestions inline, but generally nice. I like breaking sections of logic into more focused methods, and extracting constants and regex patterns into constants.

The key missing piece from this change is that there are still three points in the code where `.intercept(clientInterceptor)` and `addNettyBuilderProps(channelBuilder, properties)` are called. They should be consolidated to a single call of each, probably in the Endpoint constructor after calling the appropriate function to create the channelBuilder

bestbeforetoday commented on src/main/java/org/hyperledger/fabric/sdk/Endpoint.java:
```diff
@@ -61,243 +70,318 @@
+            );
+
+            SslContext sslContext;
+            try (InputStream myInputStream = new ByteArrayInputStream(pemBytes)) {
+                sslContext = clientContextBuilder
+                                 .trustManager(myInputStream)
+                                 .build();
```
I think this is the only point in the current code where the IOException (caught at the end of the outer try/catch block) is thrown. If so, the outer try/catch block should be removed and the code here amended as follows:
```suggestion
            try (InputStream myInputStream = new ByteArrayInputStream(pemBytes)) {
                sslContext = clientContextBuilder
                                 .trustManager(myInputStream)
                                 .build();
            } catch (IOException e) {
                throw new UncheckedIOException(e);
```

Breina committed:
Endpoint#readNtProperty -> Endpoint#readNegotionTypeProperty

Co-authored-by: Mark S. Lewis <Mark.S.Lewis@outlook.com>
Signed-off-by: Breina <info@breina.be>

Breina committed:
Endpoint#readSslpProperty -> Endpoint#readSslProviderProperty

Co-authored-by: Mark S. Lewis <Mark.S.Lewis@outlook.com>
Signed-off-by: Breina <info@breina.be>

Breina committed:
Catch IOException immediately from where it was thrown

Co-authored-by: Mark S. Lewis <Mark.S.Lewis@outlook.com>
Signed-off-by: Breina <info@breina.be>

Breina commented:
Thanks for the review! :)

Implemented all feedback.

Breina committed:
Implement code feedback

Signed-off-by: Breina <info@breina.be>

bestbeforetoday reviewed (approved):
Thank you for the contribution!
